### PR TITLE
rust ffi for scroll

### PR DESCRIPTION
**Pull request by Lawliet-Chan**
- `prove` and `verify` ffi
- Use proof and vk of `super` machine: ~/common-rs/zkevm/tmp_output_20220810_053634_nft
- Add verify cmd.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+# Install dev-packages
+RUN apt-get install libclang-dev libssl-dev llvm -y
+
+# Install Rust
+RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
+ENV PATH="/root/.cargo/bin:${PATH}"
+
+# Add Toolchain
+RUN rustup toolchain install nightly-2021-11-17
```
Comment (HAOYUatHZ) [inline Dockerfile:17]: we can use `scrolltech/rust-builder`

https://github.com/scroll-tech/roller-rs/blob/main/Dockerfile#L2
https://github.com/scroll-tech/devops/blob/main/dockerfiles/rust-builder

Comment (Lawliet-Chan) [reply Dockerfile:17]: sure

File: service/src/verify.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        .unwrap()
+        .verify_agg_circuit_proof(agg_proof)
+        .is_ok();
+    info!("verify agg-proof result: {}", verified);
+    verified as c_char
+}
+
+/// # Safety
+#[no_mangle]
+pub unsafe extern "C" fn verify_evm_proof(proof: *const c_char) -> c_char {
```
Comment (lispc) [inline service/src/verify.rs:46]: in future verify_{target_circuit like evm and state}_proof will be not used. So i think this extern "C" is not needed in fact?

File: service/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#![feature(once_cell)]
+
+pub mod prove;
+pub mod verify;
+
+pub(crate) mod utils {
+    use std::ffi::{CStr, CString};
```
Comment (lispc) [inline service/src/lib.rs:7]: CStr are truncated at '\x00'? So we must be careful here and recheck whether these codes are correct....

File: service/src/verify.rs (hunk 2 above)
Comment (Lawliet-Chan) [reply service/src/verify.rs:46]: I'm not sure if we need  target_circuit proof in the future.   Maybe we can ask ZK team?

File: service/src/lib.rs (hunk 3 above)
Comment (Lawliet-Chan) [reply service/src/lib.rs:7]: About mem leak, we free the memory in `go-roller` and `scroll` by `C.free(xxx)`. 
``CStr are truncated at '\x00'?``
I do not see it in the source code of `Cstr::from_ptr()`. You can download code and click this function to jump to the location of the source code. ( The link I cannot find it through IDEA.

File: service/src/verify.rs (hunk 2 above)
Comment (lispc) [reply service/src/verify.rs:46]: dont need. I am zk team .

Comment (Lawliet-Chan) [reply service/src/verify.rs:46]: sure

File: Makefile [hunk 4]
```diff
@@ -52,6 +52,13 @@ again:
 	MODE=sushi cargo test --features prove_verify --release test_evm_prove_verify > $(CURRENTDATE).sushi.evm.txt 2>&1; \
 	MODE=sushi cargo test --features prove_verify --release test_state_prove_verify > $(CURRENTDATE).sushi.state.txt 2>&1
 
+scroll:
+	cargo build --release
+	cp target/release/libservice.so ../scroll/verifier/lib/libverifier.so
+
+roller:
+	cargo build --release
+	cp target/release/libservice.so ../go-roller/roller/prover/lib/libprover.so
```
Comment (HAOYUatHZ) [inline Makefile:61]:
```suggestion
bin:
	cargo build --release

prepare_for_scroll: bin
	cp target/release/libservice.so ../scroll/verifier/lib/libverifier.so

prepare_for_roller: bin
	cp target/release/libservice.so ../go-roller/roller/prover/lib/libprover.so
```

Comment (HAOYUatHZ) [reply Makefile:61]: I cannot really find `target/release/libservice.so` after, `cargo build --release`, it seems this Makefile is out-of-date?

**Review by HAOYUatHZ — CHANGES_REQUESTED** [review]
Looks good, just some nitpicks.

File: Makefile (hunk 4 above)
Comment (Lawliet-Chan) [reply Makefile:61]: Yeah, it is out-of-date now. I moved `lib verifier`and `lib prover` into `scroll` and `go-roller`. I will remove these makefile cmd.

**Review by HAOYUatHZ — APPROVED** [review]

File: service/src/lib.rs (hunk 3 above)
Comment (HAOYUatHZ) [reply service/src/lib.rs]: after removing "serive" we don;t have Cstr here any more

**Pull request merged**